library/alloc/src/collections/btree/node.rs

```diff
@@ -70,66 +66,117 @@ struct LeafNode<K, V> {
     vals: [MaybeUninit<V>; CAPACITY],
 }
 
-impl<K, V> LeafNode<K, V> {
-    /// Creates a new `LeafNode`. Unsafe because all nodes should really be hidden behind
-    /// `BoxedNode`, preventing accidental dropping of uninitialized keys and values.
-    unsafe fn new() -> Self {
-        LeafNode {
-            // As a general policy, we leave fields uninitialized if they can be, as this should
-            // be both slightly faster and easier to track in Valgrind.
-            keys: MaybeUninit::uninit_array(),
-            vals: MaybeUninit::uninit_array(),
-            parent: None,
-            parent_idx: MaybeUninit::uninit(),
-            len: 0,
-        }
-    }
-}
-
-/// The underlying representation of internal nodes. As with `LeafNode`s, these should be hidden
-/// behind `BoxedNode`s to prevent dropping uninitialized keys and values. Any pointer to an
-/// `InternalNode` can be directly casted to a pointer to the underlying `LeafNode` portion of the
-/// node, allowing code to act on leaf and internal nodes generically without having to even check
+/// The actual representation of leaf or internal nodes. These should be hidden behind
+/// `BoxedNode`s to prevent dropping uninitialized keys and values. Any pointer to an
+/// internal node can be directly casted from and to a pointer to the underlying leaf portion,
+/// allowing code to act on leaf and internal nodes generically without having to even check
 /// which of the two a pointer is pointing at. This property is enabled by the use of `repr(C)`.
 #[repr(C)]
-struct InternalNode<K, V> {
-    data: LeafNode<K, V>,
+pub struct UniversalNode<K, V, E> {
+    // gdb_providers.py uses this field name for introspection.
+    leaf: LeafNode<K, V>,
 
-    /// The pointers to the children of this node. `len + 1` of these are considered
+    /// The contents of the edges delimiting each key/value pair.
+    /// For leaf nodes, these are empty. For internal nodes, these are
+    /// pointers to the children of this node. `len + 1` of these are considered
     /// initialized and valid. Although during the process of `into_iter` or `drop`,
     /// some pointers are dangling while others still need to be traversed.
-    edges: [MaybeUninit<BoxedNode<K, V>>; 2 * B],
+    edges: [MaybeUninit<E>; 2 * B],
 }
 
-impl<K, V> InternalNode<K, V> {
-    /// Creates a new `InternalNode`.
+type Node<K, V, Type> = UniversalNode<K, V, <Type as NodeTypeTrait<K, V>>::E>;
+
+// This acts as a type alias that is externally visible. The type isn't actually instantiated.
+// gdb_providers.py uses this type name for introspection.
+#[repr(C)]
```

Comment:
Seems like a good candidate for `repr(transparent)`, but this motivates the compiler to dig its heels in screaming "field is never read: `provided_node`"

Reply:
Oh, misinterpreted... simply needed a leading underscore.